synapse/federation/transport/server/_base.py

```diff
@@ -309,59 +309,65 @@ async def new_func(
                 raise
 
             # update the active opentracing span with the authenticated entity
-            set_tag("authenticated_entity", str(origin))
+            set_attribute("authenticated_entity", str(origin))
 
             # if the origin is authenticated and whitelisted, use its span context
             # as the parent.
-            context = None
+            origin_span_context = None
             if origin and whitelisted_homeserver(origin):
-                context = span_context_from_request(request)
-
-            if context:
-                servlet_span = active_span()
-                # a scope which uses the origin's context as a parent
-                processing_start_time = time.time()
-                scope = start_active_span_follows_from(
+                origin_span_context = span_context_from_request(request)
+
+            if origin_span_context:
+                local_servlet_span = get_active_span()
+                # Create a span which uses the `origin_span_context` as a parent
+                # so we can see how the incoming payload was processed while
+                # we're looking at the outgoing trace. Since the parent is set
+                # to a remote span (from the origin), it won't show up in the
+                # local trace which is why we create another span below for the
+                # local trace. A span can only have one parent so we have to
+                # create two separate ones.
+                remote_parent_span = start_active_span(
                     "incoming-federation-request",
-                    child_of=context,
-                    contexts=(servlet_span,),
-                    start_time=processing_start_time,
+                    context=origin_span_context,
+                    # Cross-link back to the local trace so we can jump
+                    # to the incoming side from the remote origin trace.
+                    links=[Link(local_servlet_span.get_span_context())],
+                )
+
+                # Create a local span to appear in the local trace
+                local_parent_span = start_active_span(
+                    "process-federation-request",
+                    # Cross-link back to the remote outgoing trace so we jump over
+                    # there.
+                    links=[Link(remote_parent_span.get_span_context())],
```

Comment:
This should be the same functionality but I found this way more straight-forward to follow (local vs remote) and why.

For more context, this was first introduced in https://github.com/matrix-org/synapse/pull/11870

Reply:
Moved to https://github.com/matrix-org/synapse/pull/13400#discussion_r935916168 so it appears on the diff